src/invoice.rs

```diff
@@ -154,7 +158,20 @@ impl FromStr for RgbInvoice {
 
         let mut chain = None;
 
-        let mut assignment = path[2].split('+');
+        let mut next_path_index = 0;
+
+        let contract = match ContractId::from_str(&path[next_path_index]) {
+            Ok(cid) => {
+                next_path_index += 1;
+                Some(cid)
+            }
+            Err(_e) => None,
+        };
```

Comment:
This opens vulnerability of introducing invalid contract names, which will be silently ignored.

I think what we need is to require that contract fragment must be always present, but may be empty meaning "any contract". In practice this should look like this: `rgb:/RGB20/6kzbKKffP6xftkxn9UP8gWqiC41W16wYKE5CYaVhmEve` (instead of `rgb:RGB20/6kzbKKffP6xftkxn9UP8gWqiC41W16wYKE5CYaVhmEve` form you are using currently in the Display and tests). With this, we skip contract if the first path segment is empty - but always fail with error if it is not containing a valid contract id.

Reply:
I've tried adding a test invoice with invalid contract ID and, without changing the code, `from_str` already fails (with imprecise error message) because since `next_path_index` doesn't get increased it tries to parse the string `RGB20` as the beneficiary, thus it returns `Err(Beneficiary("RGB20"))`.

I'm ok with requiring the contract fragment to be always present, just note that
```rust
let path = uri
    .path()
    .segments()
    .map(|e| e.to_string())
    .collect::<Vec<String>>();
```
doesn't detect the empty fragment, so string `rgb:/RGB20/6kzbKKffP6xftkxn9UP8gWqiC41W16wYKE5CYaVhmEve` and `rgb:RGB20/6kzbKKffP6xftkxn9UP8gWqiC41W16wYKE5CYaVhmEve` are indistinguishable. Therefore, in order to detect invalid contract ID and provide a precise error message, I think we'll need to manually check the invoice string to see if after `{scheme}:` there is a slash or not, I'll update PR with a proposal fix

Reply:
Pushed fix and updated tests